benchmarks/src/main/kotlin/chat/User.kt

```diff
@@ -0,0 +1,121 @@
+package chat
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.ClosedSendChannelException
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.selects.select
+import kotlinx.coroutines.yield
+import java.util.*
+import java.util.concurrent.ThreadLocalRandom
+
+/**
+ * An abstract class for server-side chat users.
+ * Every user has its own channel through which users can get messages from other users [receiveAndProcessMessage],
+ * users can send messages to their friends' channels as well [sendMessage].
+ * User contains sent and received messages metrics.
+ * To emulate real world chat servers, some work will be executed on CPU during sending and receiving messages. When user
+ * connects to the server, the connection itself consumes some CPU time.
+ * At the end of the benchmark execution [stopUser] should be called.
+ * Because of the design of the coroutines tasks scheduler and channels, it is important to call [yield] sometimes to allow other
+ * coroutines to work. This is necessary due to the fact that if a coroutine constantly has some work to do, like in this
+ * case if a coroutine has an endless flow of messages, it will work without interruption, and other coroutines will have to
+ * wait for this coroutine to end it's execution.
+ */
+abstract class User(val id: Long,
+                    val activity: Double,
+                    val messageChannel: Channel<Message>,
+                    private val configuration: BenchmarkConfiguration) {
+    var sentMessages = 0L
+        protected set
+
+    var receivedMessages = 0L
+        protected set
+
+    protected val random = Random(id)
+
+    private var messagesToSent: Double = 0.0
+
+    @Volatile
+    private var stopped = false
```

Comment:
It would be enough to have a single global flag

Reply:
It was said that it is better to create it's own flag for each user for the sake of NUMA design.

Reply:
We do not need to optimize this